### PR TITLE
Add DllMap sample

### PR DESCRIPTION
**Pull request by swaroop-sridhar**
Add a sample to demonstrate the use of NativeLibrary APIs to
map library names during resolution.

**Merge on hold for .NET Core 3.0 Preview 3**

### CODE REVIEW

File: core/extensions/DllMapDemo/Demo.cs [hunk 1]
```diff
@@ -0,0 +1,24 @@
+using System;
+using System.IO;
+using System.Reflection;
+using System.Runtime.InteropServices;
+
+public class DllMapTest
```
Comment (jkotas) [inline core/extensions/DllMapDemo/Demo.cs]: This should not be called "Test".

File: core/extensions/DllMapDemo/ReadMe.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
+# DllMap Demo
+
+This samples illustrates the use of NativeLibrary APIs to implement library-name mappings similar to [Mono](https://www.mono-project.com/) [Dllmap](http://www.mono-project.com/docs/advanced/pinvoke/dllmap/) feature.
+
+## NativeLibrary APIs
+
+.Net Core 3 provides a rich set of APIs to manage native libraries:
+
+- [NativeLibrary APIs](https://github.com/dotnet/corefx/blob/master/src/System.Runtime.InteropServices/ref/System.Runtime.InteropServices.cs#L728-L738): Perform operations on native libraries (such as `Load()`, `Free()`, get the address of an exported  symbol, etc.) in a platform-independent way from managed code.
```
Comment (jkotas) [inline core/extensions/DllMapDemo/ReadMe.md]: The links should be to do documentation, not to the source code. (e.g. https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.nativelibrary?view=netcore-3.0)

File: core/extensions/DllMapDemo/NewLib.cpp [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+#if defined(__GNUC__)
+#define EXPORT extern "C" __attribute__((visibility("default")))
+#elif defined(_MSC_VER)
+#define EXPORT extern "C" __declspec(dllexport)
+#endif 
+
+extern "C" EXPORT int NativeSum(int a, int b)
+{
+    return a + b;
+}
```
Comment (jkotas) [inline core/extensions/DllMapDemo/NewLib.cpp]: I do not think we should be committing the binary to repo. Instead, I would include a simple instructions (command line) how to build the library on Windows and Unix.

File: core/extensions/DllMapDemo/ReadMe.md (hunk 2 above)
Comment (swaroop-sridhar) [reply core/extensions/DllMapDemo/ReadMe.md]: @Jkotas The Resolver callback is not in Preview 2, so there is no documentation for it yet. 
This sample is in anticipation for Preview 3. So, I can wait for the documentation to be updated to Preview 3 and then commit this sample.

Comment (rpetrusha) [reply core/extensions/DllMapDemo/ReadMe.md]:
```suggestion
- [NativeLibrary APIs](xref:System.Runtime.InteropServices/NativeLibrary?displayProperty=nameWithType) perform operations on native libraries (such as `Load`, `Free`, get the address of an exported  symbol, etc.) in a platform-independent way from managed code.
```

File: core/extensions/DllMapDemo/ReadMe.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
+# DllMap Demo
+
+This samples illustrates the use of NativeLibrary APIs to implement library-name mappings similar to [Mono](https://www.mono-project.com/) [Dllmap](http://www.mono-project.com/docs/advanced/pinvoke/dllmap/) feature.
```
Comment (rpetrusha) [inline core/extensions/DllMapDemo/ReadMe.md]:
```suggestion
This sample illustrates the use of NativeLibrary APIs to implement library name mappings similar to the [Mono](https://www.mono-project.com/) [Dllmap](http://www.mono-project.com/docs/advanced/pinvoke/dllmap/) feature.
```

File: core/extensions/DllMapDemo/ReadMe.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+## NativeLibrary APIs
+
+.Net Core 3 provides a rich set of APIs to manage native libraries:
```
Comment (rpetrusha) [inline core/extensions/DllMapDemo/ReadMe.md]:
```suggestion
.Net Core 3.0 provides a rich set of APIs to manage native libraries:
```

File: core/extensions/DllMapDemo/ReadMe.md [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+- [NativeLibrary APIs](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.nativelibrary?view=netcore-3.0): Perform operations on native libraries (such as `Load()`, `Free()`, get the address of an exported  symbol, etc.) in a platform-independent way from managed code.
+- DllImport Resolver callback:  Get a call-back for first-chance native library resolution using custom logic.
```
Comment (rpetrusha) [inline core/extensions/DllMapDemo/ReadMe.md]:
```suggestion
- DllImport Resolver callback:  Gets a callback for first-chance native library resolution using custom logic. 
```

File: core/extensions/DllMapDemo/ReadMe.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+This samples illustrates the use of NativeLibrary APIs to implement library-name mappings similar to [Mono](https://www.mono-project.com/) [Dllmap](http://www.mono-project.com/docs/advanced/pinvoke/dllmap/) feature.
+
+## NativeLibrary APIs
+
+.Net Core 3 provides a rich set of APIs to manage native libraries:
+
+- [NativeLibrary APIs](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.nativelibrary?view=netcore-3.0): Perform operations on native libraries (such as `Load()`, `Free()`, get the address of an exported  symbol, etc.) in a platform-independent way from managed code.
+- DllImport Resolver callback:  Get a call-back for first-chance native library resolution using custom logic. 
+- [Native Library Resolve event](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.loader.assemblyloadcontext.resolvingunmanageddll?view=netcore-3.0): Get an event for last-chance native library resolution using custom logic.
```
Comment (rpetrusha) [inline core/extensions/DllMapDemo/ReadMe.md]:
```suggestion
- [Native Library Resolve event](xref:System.Runtime.Loader.AssemblyLoadContext.ResolvingUnmanagedDll?displayProperty=nameWithType>: Get an event for last-chance native library resolution using custom logic.   
```

File: core/extensions/DllMapDemo/ReadMe.md [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+.Net Core 3 provides a rich set of APIs to manage native libraries:
+
+- [NativeLibrary APIs](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.nativelibrary?view=netcore-3.0): Perform operations on native libraries (such as `Load()`, `Free()`, get the address of an exported  symbol, etc.) in a platform-independent way from managed code.
+- DllImport Resolver callback:  Get a call-back for first-chance native library resolution using custom logic. 
+- [Native Library Resolve event](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.loader.assemblyloadcontext.resolvingunmanageddll?view=netcore-3.0): Get an event for last-chance native library resolution using custom logic.   
+
+## Library Mapping
+
+The above APIs can be used to implement custom native library resolution logic, including DllMap, as illustrated in this example. The sample demonstrates:
```
Comment (rpetrusha) [inline core/extensions/DllMapDemo/ReadMe.md]:
```suggestion
These APIs can be used to implement custom native library resolution logic, including DllMap, as illustrated in this example. The sample demonstrates:
```

File: core/extensions/DllMapDemo/ReadMe.md [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+- An [app](Demo.cs) that pInvokes a method in `OldLib`, but runs in an environment where only [`NewLib`](NewLib.cpp) is available.
+- The [XML file](Demo.xml) that maps the library name from `OldLib` to `NewLib`. 
+- The [Map](Map.cs) implementation, which parses the above mapping, and uses `NativeLibrary` APIs to load the correct library.
```
Comment (rpetrusha) [inline core/extensions/DllMapDemo/ReadMe.md]:
```suggestion
- The [Map](Map.cs) implementation, which parses the above mapping and uses `NativeLibrary` APIs to load the correct library.
```

**Pull request merged**